app/services/streamlabels/index.ts

```diff
@@ -0,0 +1,212 @@
+
+    const headers = new Headers();
+    headers.append('Content-Type', 'application/json');
+
+    const url = `https://${this.hostsService.streamlabs}/api/v5/slobs/stream-labels` +
+    `/settings?token=${this.userService.widgetToken}`;
+    const request = new Request(url, {
+      method: 'POST',
+      headers,
+      body: JSON.stringify(this.settings)
+    });
+
+    return fetch(request)
+      .then(handleErrors)
+      .then(() => true);
+  }
+
+
+  /**
+   * Attempt to load initial data via HTTP instead of waiting
+   * for a socket event
+   */
+  fetchInitialData() {
+    if (!this.userService.isLoggedIn()) return;
+
+    const url = `https://${this.hostsService.streamlabs}/api/v5/slobs/stream-labels` +
+      `/files?token=${this.userService.widgetToken}`;
+    const request = new Request(url);
+
+    return fetch(request)
+      .then(handleErrors)
+      .then(response => response.json())
+      .then(json => this.setStreamlabelsData(json.data));
+  }
+
+
+  fetchSettings(): Promise<Dictionary<IStreamlabelSettings>> {
+    if (!this.userService.isLoggedIn()) return Promise.reject({});
+
+    const url = `https://${this.hostsService.streamlabs}/api/v5/slobs/stream-labels` +
+      `/settings?token=${this.userService.widgetToken}`;
+    const request = new Request(url);
+
+    return fetch(request)
+      .then(handleErrors)
+      .then(response => response.json());
+  }
+
+
+  fetchSocketToken(): Promise<string> {
+    if (!this.userService.isLoggedIn()) return Promise.reject('User must be logged in');
+
+    const url = `https://${this.hostsService.streamlabs}/api/v5/slobs/socket-token/${this.userService.widgetToken}`;
+    const request = new Request(url);
+
+    return fetch(request)
+      .then(handleErrors)
+      .then(response => response.json())
+      .then(json => json.socket_token);
+  }
+
+
+  setStreamlabelsData(data: IStreamlabelsData) {
+    this.data = { ...data };
+
+    this.subscriptions.forEach(subscription => this.updateSubscription(subscription));
+  }
+
+
+  private ensureDirectory() {
+    if (fs.existsSync(this.streamlabelsDirectory)) {
+      rimraf.sync(this.streamlabelsDirectory);
+    }
+
+    fs.mkdirSync(this.streamlabelsDirectory);
+  }
+
+
+  private get streamlabelsDirectory() {
+    return path.join(electron.remote.app.getPath('userData'), 'Streamlabels');
+  }
+
+
+  getStreamlabelsPath(filename: string) {
+    return path.join(this.streamlabelsDirectory, `${filename}.txt`);
+  }
+
+
+  /**
+   * Updates the data in a single streamlabels file
+   * @param filename the filename
+   * @param statname the name of the stat
+   */
+  updateSubscription(subscription: IStreamlabelsSubscription) {
+    if (!this.data) return;
+
+    console.log('Writing file', subscription.filename, subscription.statname);
+
+    electron.ipcRenderer.send('streamlabels-writeFile', {
+      path: this.getStreamlabelsPath(subscription.filename),
+      data: this.data[subscription.statname]
+    });
```

Comment:
Why we need main process here?

Reply:
In theory, we shouldn't need the main process at all.  Unfortunately there are a lot of issues doing asynchronous writes from the renderer process.  Synchronous writes work just fine (we do those for config saving and a few other things).  However, asynchronous writes will fail a good portion of the time.  Given the high volume of writes we need to do for streamlabels, I wasn't comfortable doing them synchronously.  Asynchronous writes work fine from the main process, so I am using an asynchronous IPC to do them from there.

As best I can tell, the issue seems to be a bug in electron.  I lost many hours to investigation and it turns out that Karl hit the exact same issue when writing the original streamlabels app (also electron).  When we have time we are going to try to reproduce in a test case and file a bug with electron.

Reply:
Ouch